jwt/jwt.go

```diff
@@ -0,0 +1,226 @@
+	if err != nil {
+		return err
+	}
+
+	block, _ := pem.Decode([]byte(publicKeyString))
+	if block == nil {
+		return errors.New("internal server error")
+	}
+
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("internal server error")
+	}
+
+	rsaPublicKey, ok := pub.(*rsa.PublicKey)
+	if !ok {
+		return errors.New("internal server error")
+	}
+
+	j.publicKey = rsaPublicKey
+	return nil
+}
+
+func getPublicKeyFromParameterStore(parameterName string) (string, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return "", err
+	}
+
+	svc := ssm.New(sess)
+	input := &ssm.GetParameterInput{
+		Name:           aws.String(parameterName),
+		WithDecryption: aws.Bool(true),
+	}
+
+	result, err := svc.GetParameter(input)
+	if err != nil {
+		return "", err
+	}
+
+	return *result.Parameter.Value, nil
+}
+
+func (j *JWTUtils) ValidateToken(tokenString string) (jwt.MapClaims, error) {
+	if j == nil || j.publicKey == nil {
+		return nil, errors.New("internal server error")
+	}
+
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("invalid token")
+		}
+		return j.publicKey, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid token")
+	}
+
+	if !token.Valid {
+		return nil, errors.New("invalid token")
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, errors.New("invalid token")
+	}
+
+	return claims, nil
+}
+
+func (j *JWTUtils) ExtractClaim(claims jwt.MapClaims, claimKey string) (string, error) {
+	if claims == nil {
+		return "", errors.New("internal server error")
+	}
+
+	value, ok := claims[claimKey].(string)
+	if !ok || value == "" {
+		return "", fmt.Errorf("unauthorized")
+	}
+	return value, nil
+}
+
+func handleMiddlewareResponse(response events.APIGatewayProxyResponse, err error) (events.APIGatewayProxyResponse, string, error) {
+	if err != nil {
+		log.Printf("JWT middleware error: %v", err)
+		return response, "", err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return response, "", nil
+	}
+
+	return response, "", nil
+}
+
+func JWTMiddleware() func(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, string, error) {
+	return func(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, string, error) {
+		jwtUtils, err := GetInstance()
+		if err != nil {
+			response := events.APIGatewayProxyResponse{
+				StatusCode: http.StatusInternalServerError,
+				Body:       "Internal server error",
+			}
+			return handleMiddlewareResponse(response, err)
+		}
+
+		cookie := req.Headers["Cookie"]
+		if cookie == "" {
+			response := events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       "Unauthenticated",
+			}
+			return handleMiddlewareResponse(response, nil)
+		}
+
+		cookieName := os.Getenv("SESSION_COOKIE_NAME")
+		if cookieName == "" {
+			env := os.Getenv("ENVIRONMENT")
```

Comment:
- instead of reading this directly from env here can we create a function that returns us all the values from the os env and we use the one required here?

Reply:
Sure can be done, 

but any benefits it will add? Why to load all the values?

Reply:
- will we not have to do `os.Getenv` with the exact key anytime we want to read a value from the env. 
- makes it easy to add/remove a value from the env as we have to add/remove it in one place (the util) and its references instead of finding by the exact key.

Reply:
Makes sense, Done